### ISSUE TITLE
Tilemap - Adding support for choosing tile draworder

### ISSUE THREAD

**Issue by TheHarvard** [godotengine/godot-proposals#664]
**Describe the project you are working on:**
2d, side scroller game with Tilemaps, using tiles supposed to overlap (to give a visual effect of perspective)

**Describe the problem or limitation you are having in your project:**
Tilemaps are drawn from top to bottom, and left to right, and this is not configurable.

This means tiles drawn later will overlap tiles drawn earlier, so a tile below another will always overlap the above tile, even when the opposite may be required to achieve a specific visual style or effect.

current workaround is to use bitmasks and a large amount of sprites with or without cutouts depending on which tile is overlapping which.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
adding support for draw order in tilemaps so that the user can configurer this.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
adding support for draw order in tilemaps so that the user can choose between:
- top to bottom, left to right
- top to bottom, right to left
- bottom to top, left to right
- bottom to top, right to left
depending on what is needed.

current implementation is: top to bottom, left to right
i personally need: top to bottom, right to left

Here is an example of the current implementation: 
(Screenshot taken from godot)
(The tiles are examples, 16X16 pixels, with textures of 32X32 pixels as an oversized example)
![Capture](https://user-images.githubusercontent.com/62014726/78364909-8f802f00-75be-11ea-8539-cfbbe407da39.PNG)

Here is a Mockup of what i need:
![Capture - Copy](https://user-images.githubusercontent.com/62014726/78365166-f7cf1080-75be-11ea-9307-9ac6661435ed.PNG)

Here is an example of the actual textures and effect i want to achieve
![Example](https://user-images.githubusercontent.com/62014726/78365579-90659080-75bf-11ea-895f-f17fbf1ac731.PNG)
:

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
as far as i am aware, you can not override the draw order of a tilemap from a script, currently. 

**Is there a reason why this should be core and not an add-on in the asset library?:**
it seems to be a relatively simple feature to a very commonly used scene, and the functionality is relevant for any project working with overlapping tiles.

**Comment by bojidar-bg** [comment]
Not to imply that the feature is not necessary, but have you tried rotating the tilemap 180⁰ degrees? This would require flipping the tiles, but it should reverse the draw order, if I'm not mistaken.

**Comment by TheHarvard** [comment]
yes, i did consider this briefly, though it would require me to reverse the tiles as you say, and set up all aspects of the tiles (like bitmasks and such) upside down in the tile editor, which would be a headache.

I'm not sure it would be a good workflow for setting up new tiles.

**Comment by TheHarvard** [comment]
As another note:

after more testing i have also found that the draw order is not always predictable.

see attached images of the same pattern of tiles, where the only difference is quadrant size 15 VS 16. 

it might be that the draw order at the quadrant edges are not predictable?
for example if the quadrants are drawn in a different order than the cells inside.

Here is of 16 Quadrant size:
(Notice the unpredictable draw order/overlap of the tiles. see specifically the top left pattern)

![Quadrant 16 ](https://user-images.githubusercontent.com/62014726/78458547-87092080-76b2-11ea-9dc9-e42241022b1c.PNG)



Here is of 15 Quadrant size:


![Quadrant 15](https://user-images.githubusercontent.com/62014726/78458550-8a041100-76b2-11ea-8df1-ee1695eb4f79.PNG)


To achieve a predictable and customizable draw order of the tiles in a tilemap then, this also will have to be investigated/addressed.

**Comment by guilhermefelipecgs** [comment]
Just to be sure, using z-index like this example https://github.com/guilhermefelipecgs/z-index_demo is not enoght for you?

**Comment by TheHarvard** [comment]
as far as i understand z-indexing, it is supposed to change which tiles are drawn in front of and behind other scene elements. 

i will have a closer look though.

**Comment by TheHarvard** [comment]
Note: I was unable to get Z-indexing to solve this issue.
